tss-esapi/src/context/tpm_commands/asymmetric_primitives.rs

```diff
@@ -78,8 +79,274 @@ impl Context {
+    /// #         tss_esapi::interface_types::algorithm::HashingAlgorithm::Sha256,
+    /// #     )
+    /// #     .expect("Failed to create session")
+    /// #     .expect("Recived invalid handle");
+    /// # let (session_attributes, session_attributes_mask) = SessionAttributesBuilder::new()
+    /// #     .with_decrypt(true)
+    /// #     .with_encrypt(true)
+    /// #     .build();
+    /// # context.tr_sess_set_attributes(session, session_attributes, session_attributes_mask)
+    /// #     .expect("Failed to set attributes on session");
+    /// # context.set_sessions((Some(session), None, None));
+    /// # let random_digest = context.get_random(16).unwrap();
+    /// # let key_auth = Auth::try_from(random_digest.value().to_vec()).unwrap();
+    /// #
+    /// // Create a key suitable for ECDH key generation
+    /// let ecc_parms = PublicEccParametersBuilder::new()
+    ///     .with_ecc_scheme(
+    ///         EccScheme::EcDh(HashScheme::new(HashingAlgorithm::Sha256)),
+    ///     )
+    ///     .with_curve(EccCurve::NistP256)
+    ///     .with_is_signing_key(false)
+    ///     .with_is_decryption_key(true)
+    ///     .with_restricted(false)
+    ///     .with_key_derivation_function_scheme(KeyDerivationFunctionScheme::Null)
+    ///     .build()
+    ///     .unwrap();
+    ///
+    /// let object_attributes = ObjectAttributesBuilder::new()
+    ///     .with_fixed_tpm(true)
+    ///     .with_fixed_parent(true)
+    ///     .with_sensitive_data_origin(true)
+    ///     .with_user_with_auth(true)
+    ///     .with_decrypt(true)
+    ///     .with_sign_encrypt(false)
+    ///     .with_restricted(false)
+    ///     .build()
+    ///     .unwrap();
+    ///
+    /// let public = PublicBuilder::new()
+    ///     .with_public_algorithm(PublicAlgorithm::Ecc)
+    ///     .with_name_hashing_algorithm(HashingAlgorithm::Sha256)
+    ///     .with_object_attributes(object_attributes)
+    ///     .with_ecc_parameters(ecc_parms)
+    ///     .with_ecc_unique_identifier(&EccPoint::default())
+    ///     .build()
+    ///     .unwrap();
+    ///
+    /// let key_handle = context
+    ///     .create_primary(
+    ///         Hierarchy::Owner,
+    ///         &public,
+    ///         Some(&key_auth),
+    ///         None,
+    ///         None,
+    ///         None,
+    ///     )
+    ///     .unwrap()
+    ///     .key_handle;
+    ///
+    /// // Generate ephemeral key pair and a shared secret
+    /// let (z_point, pub_point) = context.ecdh_key_gen(key_handle).unwrap();
+    /// let z_point_gen = context.ecdh_z_gen(key_handle, pub_point).unwrap();
+    /// assert_eq!(z_point.x().value(), z_point_gen.x().value());
+    /// ```
+    pub fn ecdh_z_gen(&mut self, key_handle: KeyHandle, in_point: EccPoint) -> Result<EccPoint> {
+        let mut point = null_mut();
+        let in_point: TPMS_ECC_POINT = in_point.into();
+        let in_point = TPM2B_ECC_POINT {
+            size: in_point.x.size + in_point.y.size,
```

Comment:
@Superhepper @ionut-arm folks, could you take a look if this is a correct way to convert `TPMS_ECC_POINT` into `TPM2B_ECC_POINT`? Thanks!

Reply:
I'm not sure I follow exactly what they mean in the spec by
> if  the  point  is  to  be  omitted,  the  X  and  Y  coordinates  need  to  be  individually  set  to  Empty  Buffers.  The 
minimum value for size will be four. It is checked indirectly by unmarshaling of the TPMS_ECC_POINT. If 
the type of point were BYTE, then size could have been zero. However, this would complicate the process 
of marshaling the structure. 

As in, I'm unsure what you'd put in `TPMS_ECC_POINT` if you wanted to have the `size` in `TPM2B_ECC_POINT` as 4 - do you just set everything to 0 (as in, just `Default::default`)?
The "normal" value for `size` should be `::std::mem::size_of::<TPMS_ECC_POINT>()`. I think your calculation only covers the sizes of the buffers, but not of the whole structure (since they each contain a `size` field as well)

Reply:
hrrm... I am not sure. I will look into it. I think it would be better if the conversion into TPM2B_ECC_POINT is added to EccPoint struct.

Reply:
> I think it would be better if the conversion into TPM2B_ECC_POINT is added to EccPoint struct.

Yes, agreed!

> hrrm... I am not sure. I will look into it.

Greatly appreciated :bow:

Reply:
As long as you do not want the parameter to be encrypted then it is fine to use size = 0 but that is not something we can do. We need to support encryption of parameters so the calculation of the size should be:

```rust
size = std::mem::size_of::<u16>() + point.x.len() + std::mem::size_of::<u16>() + point.y.len()
```

I think ....

Reply:
> As long as you do not want the parameter to be encrypted then it is fine to use size = 0 

I think that's what the spec tries to explain away in what I commented above. Have you tried using it with either @Superhepper's equation or `::std::mem::size_of::<TPMS_ECC_POINT>()`?

Reply:
@ionut-arm I do not think ```::std::mem::size_of::<TPMS_ECC_POINT>()``` because that will suggest that the entire allocated space contains valid data. And in the case where a point is to be omitted the x and y needs to be empty buffers but still the marshalling needs to work so the size cannot be 0 it needs to be four so that the size parameters of the TPMS_ECC_POINT can be read and confirmed that they are empty buffers.

Reply:
Ok, but if you look at what the description of the `size` field says: `size of the remainder of this structure`. And the remainder of the structure is a `TPMS_ECC_POINT`, which at least in Rust _cannot_ fit in 4 bytes. I assume the TPM just takes the `size: 4` as an "edge" case and knows to simply assume it's empty. So maybe in the `EccPoint` => `TPM2B_ECC_POINT` conversion we can have an edge case of our own if both `x` and `y` are empty and set it to 4, otherwise just use the size of `TPMS_ECC_POINT`?

Reply:
Yes, wording in the specification made me seriously confused because I didn't understand what they meant with ```remainder``` but I interpret it as data left in the structure that is valid. So when came up with the equation so I thought that the size field in the ```TPM2B_ECC_PARAMETER``` is the only thing that always contains valid data i.e. ```std::mem::size_of::<u16>()``` and there is one such field for x and one for y so the minimum number of valid bytes must be four. So the equation should work for the empty buffer case as well.

like this:
```rust
TPM2B_ECC_POINT {
    size: u16,
    point: TPM2_ECC_POINT {
        x: TPM2B_ECC_PARAMETER {
            size: u16 // 2 Bytes valid data always
            buffer: [u8; MAX_ECC_KEY_BYTES] // Contains 0 to MAX_ECC_KEY_BYTES valid bytes
        }
        y: TPM2B_ECC_PARAMETER {
            size: u16 // 2 Bytes valid data always
            buffer: [u8; MAX_ECC_KEY_BYTES] // Contains 0 to MAX_ECC_KEY_BYTES valid bytes
        }        
    },
}
```

Reply:
Thanks for the elaborate explanation of what's happening with the `size` field folks! I've adjusted the code moving conversion to `TPM2B_ECC_POINT` into a proper `From` impl and reused that in the ECDH function using @Superhepper's equation.

I think this is ready for a next round of reviews. Thanks in advance!

Reply:
@Superhepper - do you want to have another look at the `From` implementation?